File: backend/app/initiatives/calculations.py
from datetime import date, datetime

from app.initiatives.models import CalculatedMetrics


def _to_float(value) -> float:
    try:
        return float(value or 0)
    except (TypeError, ValueError):
        return 0.0


def _months_since(dt) -> float:
    """Meses decorridos desde uma data ate hoje."""
    if not dt:
        return 0.0
    if isinstance(dt, str):
        try:
            dt = datetime.fromisoformat(dt.replace("Z", "+00:00"))
        except Exception:
            return 0.0
    if isinstance(dt, datetime):
        dt = dt.date()
    delta = date.today() - dt
    return max(0.0, delta.days / 30.44)


def calculate_metrics(data: dict) -> CalculatedMetrics:
    """
    Formula central:

    CAPEX = (horas_dev * custo_hora_dev)
          + (horas_devops * custo_hora_devops)
          + (horas_terceiros * custo_hora_terceiros)

    OPEX = ganhos mensais operacionais - custos mensais recorrentes

    DevOps entra apenas como investimento one-time no CAPEX da iniciativa.
    """

    time_saved_per_day = _to_float(data.get("time_saved_per_day"))
    execution_days_per_month = _to_float(data.get("execution_days_per_month"))
    affected_people_count = _to_float(data.get("affected_people_count"))
    cost_per_hour = _to_float(data.get("cost_per_hour"))

    hours_per_person = time_saved_per_day * execution_days_per_month
    total_hours_saved = hours_per_person * affected_people_count
    gain_hours = total_hours_saved * cost_per_hour

    headcount_reduction = _to_float(data.get("headcount_reduction"))
    monthly_employee_cost = _to_float(data.get("monthly_employee_cost"))
    productivity_increase = _to_float(data.get("productivity_increase"))
    additional_task_value = _to_float(data.get("additional_task_value"))

    gain_headcount = headcount_reduction * monthly_employee_cost
    gain_productivity = productivity_increase * additional_task_value
    total_gains_opex = gain_hours + gain_headcount + gain_productivity

    maintenance_hours = _to_float(data.get("maintenance_hours"))
    tech_hour_cost = _to_float(data.get("tech_hour_cost"))
    token_cost = _to_float(data.get("token_cost"))
    cloud_infra_cost = _to_float(data.get("cloud_infra_cost"))
    maintenance_cost_monthly = (maintenance_hours * tech_hour_cost) + token_cost + cloud_infra_cost
    net_gains_monthly = total_gains_opex - maintenance_cost_monthly

    development_estimate_seconds = _to_float(data.get("development_estimate_seconds"))
    development_estimate_hours = development_estimate_seconds / 3600
    capex_dev = development_estimate_hours * tech_hour_cost

    devops_hours = _to_float(data.get("devops_hours"))
    devops_hour_cost = _to_float(data.get("devops_hour_cost"))
    capex_devops = devops_hours * devops_hour_cost

    third_party_hours = _to_float(data.get("third_party_hours"))
    third_party_hour_cost = _to_float(data.get("third_party_hour_cost"))
    capex_third_party = third_party_hours * third_party_hour_cost

    total_capex = capex_dev + capex_devops + capex_third_party

    time_spent_seconds = _to_float(data.get("time_spent_seconds"))
    time_spent_hours = time_spent_seconds / 3600

    time_variance_percent = None
    if development_estimate_hours > 0 and time_spent_hours > 0:
        time_variance_percent = round(
            ((time_spent_hours - development_estimate_hours) / development_estimate_hours) * 100,
            1,
        )

    roi_percent = None
    if total_capex > 0:
        roi_percent = round((net_gains_monthly / total_capex) * 100, 2)

    roi_percent_real = None
    if time_spent_hours > 0 and development_estimate_hours > 0:
        capex_real = (time_spent_hours * tech_hour_cost) + capex_devops + capex_third_party
        if capex_real > 0:
            roi_percent_real = round((net_gains_monthly / capex_real) * 100, 2)

    roi_accumulated = None
    resolution_date = data.get("resolution_date")
    completion_date = resolution_date or data.get("status_updated_at")
    months_live = _months_since(completion_date)

    if time_spent_hours > 0:
        capex_for_accumulated = (time_spent_hours * tech_hour_cost) + capex_devops + capex_third_party
    else:
        capex_for_accumulated = total_capex

    if capex_for_accumulated > 0 and months_live > 0:
        total_net_gains_so_far = net_gains_monthly * months_live
        roi_accumulated = round(
            ((total_net_gains_so_far - capex_for_accumulated) / capex_for_accumulated) * 100,
            2,
        )
    elif capex_for_accumulated > 0 and months_live == 0 and completion_date:
        roi_accumulated = -100.0

    payback_months = None
    if net_gains_monthly > 0:
        capex_for_payback = capex_for_accumulated if capex_for_accumulated else total_capex
        payback_months = round(capex_for_payback / net_gains_monthly, 2)

    return CalculatedMetrics(
        total_gains=round(net_gains_monthly, 2),
        total_costs=round(total_capex, 2),
        roi_percent=roi_percent,
        roi_percent_real=roi_percent_real,
        roi_accumulated=roi_accumulated,
        months_live=round(months_live, 1) if months_live > 0 else (0.0 if completion_date else None),
        total_hours_saved=round(total_hours_saved, 1),
        payback_months=payback_months,
        development_estimate_hours=round(development_estimate_hours, 2),
        time_spent_hours=round(time_spent_hours, 2),
        time_variance_percent=time_variance_percent,
        capex_development_cost=round(capex_dev, 2),
        capex_devops_cost=round(capex_devops, 2),
        capex_third_party_cost=round(capex_third_party, 2),
    )

File: frontend/src/utils/calculations.js
/**
 * Formula canonica de metricas no frontend.
 * Alinhada com backend/app/initiatives/calculations.py.
 */
export function calculateMetrics(data) {
  const timeSavedPerDay = Number(data.time_saved_per_day || 0)
  const executionDaysPerMonth = Number(data.execution_days_per_month || 0)
  const affectedPeopleCount = Number(data.affected_people_count || 0)
  const costPerHour = Number(data.cost_per_hour || 0)

  const totalHoursSaved = timeSavedPerDay * executionDaysPerMonth * affectedPeopleCount
  const gainHours = totalHoursSaved * costPerHour

  const gainHeadcount = Number(data.headcount_reduction || 0) * Number(data.monthly_employee_cost || 0)
  const gainProductivity = Number(data.productivity_increase || 0) * Number(data.additional_task_value || 0)
  const totalGainsOpex = gainHours + gainHeadcount + gainProductivity

  const techHourCost = Number(data.tech_hour_cost || 0)
  const maintenanceCost =
    Number(data.maintenance_hours || 0) * techHourCost +
    Number(data.token_cost || 0) +
    Number(data.cloud_infra_cost || 0)
  const netMonthlyGain = totalGainsOpex - maintenanceCost

  const developmentEstimateHours = Number(data.development_estimate_seconds || 0) / 3600
  const timeSpentHours = Number(data.time_spent_seconds || 0) / 3600

  const capexDev = developmentEstimateHours * techHourCost
  const capexDevops = Number(data.devops_hours || 0) * Number(data.devops_hour_cost || 0)
  const capexThirdParty = Number(data.third_party_hours || 0) * Number(data.third_party_hour_cost || 0)
  const initialInvestment = capexDev + capexDevops + capexThirdParty

  const roiPercent = initialInvestment > 0
    ? Math.round((netMonthlyGain / initialInvestment) * 10000) / 100
    : null

  let roiPercentReal = null
  if (timeSpentHours > 0 && developmentEstimateHours > 0) {
    const capexReal = timeSpentHours * techHourCost + capexDevops + capexThirdParty
    if (capexReal > 0) {
      roiPercentReal = Math.round((netMonthlyGain / capexReal) * 10000) / 100
    }
  }

  let roiAccumulated = null
  let monthsLive = null
  const completionDate = data.resolution_date || data.status_updated_at
  if (completionDate) {
    const diffMs = Date.now() - new Date(completionDate).getTime()
    monthsLive = Math.max(0, diffMs / (1000 * 60 * 60 * 24 * 30.44))

    const capexForAccumulated = timeSpentHours > 0
      ? timeSpentHours * techHourCost + capexDevops + capexThirdParty
      : initialInvestment

    if (capexForAccumulated > 0) {
      const accumulatedNetGain = netMonthlyGain * monthsLive
      roiAccumulated = Math.round(((accumulatedNetGain - capexForAccumulated) / capexForAccumulated) * 10000) / 100
    }
  }

  let paybackMonths = null
  if (netMonthlyGain > 0) {
    const capexForPayback = timeSpentHours > 0
      ? timeSpentHours * techHourCost + capexDevops + capexThirdParty
      : initialInvestment
    paybackMonths = Math.round((capexForPayback / netMonthlyGain) * 100) / 100
  }

  let timeVariancePercent = null
  if (developmentEstimateHours > 0 && timeSpentHours > 0) {
    timeVariancePercent = Math.round(
      ((timeSpentHours - developmentEstimateHours) / developmentEstimateHours) * 1000
    ) / 10
  }

  return {
    total_gains: Math.round(netMonthlyGain * 100) / 100,
    total_costs: Math.round(initialInvestment * 100) / 100,
    roi_percent: roiPercent,
    roi_percent_real: roiPercentReal,
    roi_accumulated: roiAccumulated,
    months_live: monthsLive != null ? Math.round(monthsLive * 10) / 10 : (completionDate ? 0 : null),
    total_hours_saved: Math.round(totalHoursSaved * 10) / 10,
    payback_months: paybackMonths,
    development_estimate_hours: Math.round(developmentEstimateHours * 100) / 100,
    time_spent_hours: Math.round(timeSpentHours * 100) / 100,
    time_variance_percent: timeVariancePercent,
    capex_development_cost: Math.round(capexDev * 100) / 100,
    capex_devops_cost: Math.round(capexDevops * 100) / 100,
    capex_third_party_cost: Math.round(capexThirdParty * 100) / 100,
  }
}
